Make the glue compute short circuit only if possible

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94512
* #94473
* __->__ #94437
* #94431

If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Make the glue compute short circuit only if possible

If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

bdhirsh commented on c10/core/TensorImpl.cpp:
```diff
@@ -560,56 +560,82 @@ SymBool TensorImpl::compute_non_overlapping_and_dense(identity<SymBool>) const {
+    return true;
+  }
+  return extra_meta_->is_contiguous_ |
+      extra_meta_->is_channels_last_contiguous_ |
```
hmm is there a reason for the change from logical or to bitwise or? (Do we not want to use || on SymBools?)

bdhirsh replied:
Oh, from looking through the next PR, it looks like our python sympy impl eventually calls `eval_is_non_overlapping_and_dense()`, which treats the output as an int.

ezyang replied:
Logical or is not overrideable, bitwise or goes into `sym_or` which is branchless.

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make the glue compute short circuit only if possible"


If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
